lib/mongo/retryable.rb

```diff
@@ -98,12 +98,58 @@ def read_with_one_retry
     # @return [ Result ] The result of the operation.
     #
     # @since 2.1.0
-    def write_with_retry(session, server_selector)
+    def write_with_retry(session, write_concern, &block)
+      unless retry_write_allowed?(session, write_concern)
+        return legacy_write_with_retry(&block)
+      end
+
+      server = cluster.next_primary
+      unless server.retry_writes?
+        return legacy_write_with_retry(server, &block)
+      end
+
+      begin
+        txn_num = session.next_txn_num
+        yield(server, txn_num)
+      rescue Error::SocketError, Error::SocketTimeoutError => e
+        retry_write(e, txn_num, &block)
+      rescue Error::OperationFailure => e
+        raise e unless e.write_retryable?
+        retry_write(e, txn_num, &block)
+      end
+    end
+
+    private
+
+    def retry_write_allowed?(session, write_concern)
+      session && session.retry_writes? &&
+          (write_concern.nil? || write_concern.acknowledged?)
+    end
+
+    def retry_write(original_error, txn_num, &block)
+      log_retry(original_error)
+      cluster.scan!
+      server = cluster.next_primary
+      raise original_error unless (server.retry_writes? && txn_num)
+      yield(server, txn_num)
+    rescue Error::SocketError, Error::SocketTimeoutError => e
+      cluster.scan!
+      raise e
+    rescue Error::OperationFailure => e
+      raise original_error unless e.write_retryable?
+      cluster.scan!
+      raise e
+    rescue
+      raise original_error
+    end
+
+    def legacy_write_with_retry(server = nil)
       attempt = 0
       begin
         attempt += 1
-        yield(server_selector.call)
+        yield(server || cluster.next_primary)
       rescue Error::OperationFailure => e
+        server = nil
         raise(e) if attempt > Cluster::MAX_WRITE_RETRIES
```

Comment:
If I understand the purpose of this line correctly, it's to abstract away the exact value of MAX_WRITE_RETRIES instead of assuming some value (e.g. 1). If that's the case, should this be in a loop? If MAX_WRITE_RETRIES is greater than 1, this function will still only retry once.

Reply:
Can you tell me why you think it'll only retry once? If {{attempt}} is less than {{Cluster::MAX_WRITE_RETRIES}} then the error will not be raised. Then, if the error is not {{write_retryable?}}, execution will continue until line 158, when the block will be retried.
Keep in mind that this is the old behavior before we wrote the retryable writes spec. It's still used for methods like update_many and delete_many and writes to servers < 3.6.)

Reply:
Ah, the Github diff viewer had hidden the part with the retry, and I didn't notice that the were lines missing. Sorry about that!

Reply:
oh, that explains it! No worries